contracts/actions/llamalend/LlamaLendSelfLiquidate.sol

```diff
@@ -0,0 +1,109 @@
+        emit ActionEvent("LlamaLendSelfLiquidate", logData);
+        return bytes32(amountPulled);
+    }
+
+    /// @inheritdoc ActionBase
+    function executeActionDirect(bytes memory _callData) public payable virtual override {
+        Params memory params = parseInputs(_callData);
+
+        (, bytes memory logData) = _llamaLendSelfLiquidate(params);
+        logger.logActionDirectEvent("LlamaLendSelfLiquidate", logData);
+    }
+
+    /// @inheritdoc ActionBase
+    function actionType() public pure virtual override returns (uint8) {
+        return uint8(ActionType.STANDARD_ACTION);
+    }
+
+    //////////////////////////// ACTION LOGIC ////////////////////////////
+
+    function _llamaLendSelfLiquidate(Params memory _params) internal returns (uint256, bytes memory) {
+
+        uint256 userWholeDebt = ILlamaLendController(_params.controllerAddress).debt(address(this));
+        (uint256 collInDebtAsset, uint256 collInCollateralAsset) = getCollAmountsFromAMM(_params.controllerAddress, address(this));
+
+        uint256 amountToPull;
+        address debtAsset = ILlamaLendController(_params.controllerAddress).borrowed_token();
+        
+        if (collInDebtAsset < userWholeDebt) {
+            /// @dev in some cases debt - collInCrvUsd will fall few wei short of closing the position
+            // if we don't have enough crvUsd in coll, pull the rest from the user
+            amountToPull = userWholeDebt - collInDebtAsset + EXTRA_BUFFER;
+            amountToPull = debtAsset.pullTokensIfNeeded(_params.from, amountToPull);
+            debtAsset.approveToken(_params.controllerAddress, amountToPull);
+        }
+
+        address collateralAsset = ILlamaLendController(_params.controllerAddress).collateral_token();
+        
+        uint256 collAssetBalancePreLiq = collateralAsset.getBalance(address(this));
+        uint256 debtAssetBalancePreLiq = debtAsset.getBalance(address(this));
+
+        ILlamaLendController(_params.controllerAddress).liquidate(address(this), _params.minDebtAssetExpected, false);
+
+        uint256 collAssetBalanceAfterLiq = collateralAsset.getBalance(address(this));
+
+        uint256 collAssetReceivedFromLiq = collAssetBalanceAfterLiq - collAssetBalancePreLiq;
+        collateralAsset.withdrawTokens(_params.to, collAssetReceivedFromLiq);
+
+        uint256 debtAssetBalanceAfterLiq = debtAsset.getBalance(address(this));
+        
+        if (collInDebtAsset > userWholeDebt) {
+            // we return any extra debt asset that was left in coll after liquidation
+            debtAsset.withdrawTokens(_params.to, debtAssetBalanceAfterLiq - debtAssetBalancePreLiq);
+        } else {
+            // we return any extra debt asset that was not needed in debt and remove any extra approval
+            debtAsset.withdrawTokens(_params.from, amountToPull - (debtAssetBalancePreLiq - debtAssetBalanceAfterLiq));
+            IERC20(debtAsset).approve(_params.controllerAddress, 0);
```

Comment:
Doesn't make much sense since it would just use up more gas if approve 0 were to fail, since SafeERC20 would try approving 0, and if it fails try approving 0 again without any changes